runtime/jvmti/suspendhelper.cpp

```diff
@@ -45,7 +45,15 @@ suspendThread(J9VMThread *currentThread, jthread thread, UDATA allowNull, UDATA
 				if (currentThread == targetThread) {
 					*currentThreadSuspended = TRUE;
 				} else {
-					currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(currentThread);
+					J9InternalVMFunctions *vmFuncs = currentThread->javaVM->internalVMFunctions;
+					/* It is a well known problem that suspending a Java thread may cause deadlocks in the application if a resource is 
+					 * held while suspended. To prevent this happening in the Thread JCL library, check that the Object java.lang.Thread.lock 
+					 * has been released before suspending. To make sure of this, acquire the lock while halt is in progress.
+					 */
+					vmFuncs->objectMonitorEnter(currentThread, J9VMJAVALANGTHREAD_LOCK(targetThread, targetThread->threadObject));
```

Comment:
One final thing - the object field access macros should take `currentThread` not `targetThread`.

Reply:
```
vmFuncs->objectMonitorEnter(currentThread, J9VMJAVALANGTHREAD_LOCK(currentThread, currentThread->threadObject));
```

Wouldn't that make it so `currentThread` is acquiring the lock that belongs to its own class?

The way I'm reading the existing line is `currentThread` is acquiring `targetThread`'s lock to make sure that no `targetThread` JCL methods are holding it.

Reply:
Only the first parameter should be changed - it's used in some GC policies and must be the current thread.

Reply:
Still waiting for this to be fixed before test/merge.